### ISSUE TITLE
Consider enforcing no leading underscore in tag keys

### ISSUE THREAD

**Issue by e-dard** [influxdata/influxdb#11961]
I would like to consider ensuring that the storage engine does not accept any tag keys with leading underscores, essentially leaving the `_` prefix as a private key space for internal series key information. 

With that as an invariant, we could make certain performance improvements around reading and parsing keys.

**Comment by e-dard** [comment]
This was resolved via https://github.com/influxdata/influxdb/pull/12391

**Comment by zhulongcheng** [comment]
`measurement` overwrite `_measurement` tag in flux table.

```
// write
any,_field=ff,_measurement=mm,_tag=tt foo="bar"

// query: "from(bucket:"purple")|> range(start:-10h)"
,result,table,_start,_stop,_time,_value,_field,_measurement,_tag
,_result,1,2019-05-31T08:09:08.867394Z,2019-05-31T08:19:08.867394Z,2019-05-31T08:19:00Z,bar,foo,any,tt


```

at least, `_start,_stop,_time,_value,_field,_measurement` these tag keys should not be accepted.

**Comment by zhulongcheng** [comment]
https://github.com/influxdata/influxdb/blob/530ceb9cbdae077302f37201bf3ef6cb8192e4fc/storage/readservice/cursor.go#L143-L153

**Comment by e-dard** [comment]
@zhulongcheng thanks for the issue! Hmm I think this is an issue in the read side. @stuartcarnie can you take a look?

**Comment by stuartcarnie** [comment]
@e-dard the issue I see here is that `_measurement` was allowed via a write, and therefore stored as a real tag. Specifically, the series key is stored as:

```
<org><bucket>,\x00=any,_field=ff,_measurement=mm,_tag=tt,\xff=foo
```

which should never be allowed.

When a read occurs, the `\x00` and `\xff` bytes are remapped to `_measurement` and `_field` respectively:

https://github.com/influxdata/influxdb/blob/530ceb9cbdae077302f37201bf3ef6cb8192e4fc/storage/readservice/cursor.go#L143-L153

**Comment by e-dard** [comment]
@stuartcarnie yeah seems about right. I think then we’ll have to check and block these. I’ll put this on the back log.

**Comment by zhulongcheng** [comment]
don't close.
